### PR TITLE
gh-110019: Refactor summarize_stats

### PR DESCRIPTION
**Pull request by mdboom**
This refactors summarize_stats so that the comparative tables are easier to make and use more common code.

Reviewing this as a diff may be rather difficult -- instead maybe just look at the file verbatim.

* Issue: gh-110019

### CODE REVIEW

File: Tools/scripts/summarize_stats.py [hunk 1]
```diff
@@ -2,908 +2,988 @@
...
-    common key.
-    """
-    if len(a_rows) == 0 and len(b_rows) == 0:
-        return []
 
-    if len(a_rows):
-        a_ncols = list(set(len(x) for x in a_rows))
-        if len(a_ncols) != 1:
-            raise ValueError("Table a is ragged")
+class Stats(dict):
```
Comment (markshannon) [inline Tools/scripts/summarize_stats.py]: Inheriting from builtin collections can be awkward.
Could you wrap the dict?

File: Tools/scripts/summarize_stats.py [hunk 2]
```diff
@@ -2,908 +2,988 @@
...
+                int(k): v for k, v in self["_stats_defines"].items()
+            }
+            self["_defines"] = {int(k): v for k, v in self["_defines"].items()}
 
-    if len(a_rows):
-        ncols = a_ncols[0]
-    else:
-        ncols = b_ncols[0]
+        elif input.is_dir():
+            stats: collections.Counter = collections.Counter()
```
Comment (markshannon) [inline Tools/scripts/summarize_stats.py]: This type annotation seems redundant

File: Tools/scripts/summarize_stats.py [hunk 3]
```diff
@@ -2,908 +2,988 @@
...
+            if "__" not in op
+        ]
+        self["_stats_defines"] = get_defines(
+            Path("Include") / "cpython" / "pystats.h", "EVAL_CALL"
         )
-    return rows
+        self["_defines"] = get_defines(Path("Python") / "specialize.c")
 
+    @property
+    def defines(self) -> Defines:
```
Comment (markshannon) [inline Tools/scripts/summarize_stats.py]: If you wrap the dict, then this can be a normal attribute.

File: Tools/scripts/summarize_stats.py [hunk 4]
```diff
@@ -2,908 +2,988 @@
...
+        return format(self, ",d")
 
-        import opcode
 
-        stats["_specialized_instructions"] = [
-            op for op in opcode._specialized_opmap.keys() if "__" not in op
-        ]
-        stats["_stats_defines"] = get_stats_defines()
-        stats["_defines"] = get_defines()
+class CountPer:
```
Comment (markshannon) [inline Tools/scripts/summarize_stats.py]: Why not use a `Ratio`, or just a `float`?

File: Tools/scripts/summarize_stats.py [hunk 5]
```diff
@@ -2,908 +2,988 @@
...
+class Section:
+    def __init__(
+        self,
+        title: str,
+        summary: str = "",
+        part_iter=None,
+        comparative: bool = True,
+    ):
+        self.title = title
+        if not summary:
```
Comment (markshannon) [inline Tools/scripts/summarize_stats.py]: If the summary is explicitly "", it is ignored. Keeping the default as `None` seems better.

File: Tools/scripts/summarize_stats.py [hunk 6]
```diff
@@ -2,908 +2,988 @@
...
+        print(header_line, file=out)
+        print(under_line, file=out)
+        for row in rows:
+            if width is not None and len(row) != width:
+                raise ValueError("Wrong number of elements in row '" + str(row) + "'")
+            print("|", " | ".join(to_markdown(i) for i in row), "|", file=out)
+        print(file=out)
+
+
+class FixedTable(Table):
```
Comment (markshannon) [inline Tools/scripts/summarize_stats.py]: I don't like all these subclasses of `Table`, it mixes up the responsibilities of creating the contents and the formatting.
Having a single `Table` class which does the formatting and takes the contents as a parameter to its `__init__` would separate the responsibilities better. 
So instead of `ExecutionCountTable("uops")`, it would be `Table("uops", get_uop_counts())`

**Review by markshannon — COMMENTED** [review]
Thanks for this. This file definitely needed refactoring.

I have few comments. I think a more functional (as opposed to OO) style would help clarity, but the general design seems sound.

File: Tools/scripts/summarize_stats.py (hunk 2 above)
Comment (mdboom) [reply Tools/scripts/summarize_stats.py]: Yep, but mypy requires it :(

File: Tools/scripts/summarize_stats.py (hunk 4 above)
Comment (mdboom) [reply Tools/scripts/summarize_stats.py]: Because my secret plan is to also introduce CSV output, where we would want to format this differently.

File: Tools/scripts/summarize_stats.py (hunk 6 above)
Comment (mdboom) [reply Tools/scripts/summarize_stats.py]: It's more complicated than that.  Every table knows how to generate a single set of results, and then also supports combining two tables for comparative results.  This usually is straightforward, but for some tables (e.g. execution count table) that behavior needs to be overridden.  But I'll take a look at doing all of this in a more functional way -- I'm a little worried we'll end up back where we started, though.

File: Tools/scripts/summarize_stats.py (hunk 4 above)
Comment (mdboom) [reply Tools/scripts/summarize_stats.py]: And also a Ratio is rendered as a percentage, a CountPer is rendered as an integer.  "Uops run per trace" is much better represented as an integer rather than a percentage.

File: Tools/scripts/summarize_stats.py (hunk 2 above)
Comment (markshannon) [reply Tools/scripts/summarize_stats.py]: Don't use mypy then?

Is there a Mypy issue for this?

Comment (AlexWaygood) [reply Tools/scripts/summarize_stats.py]: This should make mypy happy and Mark less unhappy:

```suggestion
            stats = collections.Counter[str]()
```

File: Tools/scripts/summarize_stats.py (hunk 4 above)
Comment (markshannon) [reply Tools/scripts/summarize_stats.py]: I think uops per trace is more naturally a float than an int. I agree that 30[.0] is better than 3000% though.
Maybe add a `percent: bool=True` argument to `Ratio`'s `__init__`?

File: Tools/scripts/summarize_stats.py (hunk 6 above)
Comment (markshannon) [reply Tools/scripts/summarize_stats.py]: The data can support merging, etc. `get_uop_counts()` can return an object that supports that functionality.
I just think that separating formating from data manipulation will make things more maintainable.

File: Tools/scripts/summarize_stats.py (hunk 3 above)
Comment (mdboom) [reply Tools/scripts/summarize_stats.py]: Yes, but since this value should be saved, it's convenient for it to be on the dictionary which is dumped to JSON.

File: Tools/scripts/summarize_stats.py (hunk 2 above)
Comment (AlexWaygood) [reply Tools/scripts/summarize_stats.py]: > Don't use mypy then?
> 
> Is there a Mypy issue for this?

It's not a mypy bug. Pyright will complain at you in exactly the same way about this. Mypy can't tell what kind of items are going to be stored as keys for the `Counter`, so it demands an explicit annotation. @mdboom's annotation shuts mypy up, because `collections.Counter` as a type annotation is equivalent to `collections.Counter[Any]`. But the better solution is to use `collections.Counter[str]`, because they keys of `stats` are all strings.

Comment (markshannon) [reply Tools/scripts/summarize_stats.py]: It is a mypy bug.
Omitting the annotation and providing `collections.Counter` as an annotation has exactly the same information content. So complaining about one and not the other is erroneous.
I assume mypy will not complain about `stats = collections.Counter[str]()` then.

Comment (AlexWaygood) [reply Tools/scripts/summarize_stats.py]: I agree there's a usability bug here; mypy isn't communicating what it wants from you very clearly at all.

> I assume mypy will not complain about `stats = collections.Counter[str]()` then.

Correct, that will make mypy happy (https://github.com/python/cpython/pull/110398#discussion_r1347626014)

**Comment by mdboom** [comment]
I have modified this to:

1) Take a more functional approach without inheritance
2) Separate out markdown output from data handling
3) Not have `Stats` inherit from `dict`

**Comment by markshannon** [comment]
I think what bothers me with this PR is that the data processing is mixed in with the data storage. This can be a problem with OOP.

There is nothing wrong with objects that enhance and structure the data, but data processing should be separated, otherwise the code can be hard to follow.

I think we need the following:
1. To read the data off disk and into one big poorly structured object ("blob") (we already have this  https://github.com/python/cpython/blob/main/Tools/scripts/summarize_stats.py#L229-L244)
2. A structured data type, `Stats`.
3. A function to convert the "blob" to `Stats`
4. Functions/methods to save and load the `Stats` to a file (currently json)
5. A function/method to make a single `Stats` from the diff of two `Stats`
6. A function to output `Stats` to a human readable format (currently markdown)

Each "function" above doesn't have to one function, but should be independent of the others.

By all means use objects (this is Python, not Haskell), but I'd recommend trying to stick to functional(ish) principles:
* All classes should have a simple `__init__` (as would be generated for a `dataclass`)
* No method should mutate the object
* No special methods
* Methods should stick to the domain of the object. So for a `Section` object, `to_rows()->list[tuple[str]]` is fine, but `write_markdown()` maybe not so much.

With that, the base pipeline (raw stats files to markdown) would look something like:

```Py
def main():
    # Process args and find folders, etc.
    raw_stats = gather_stats(stats_dir)
    stats = structure_stats(raw_stats)
    output_stats(stats, outfile)
```

The `Stats` class should be a structured data type, with the top level containing attributes for each of the top level categories in the data, `execution_counts`, `pair_counts`, `predecessor_pairs`, etc.
It could simply be a list of `Section`s, where the Sections describe the data, if that makes more sense. The markdown file is structured as a list of sections.

A `Section` will need to describe how to present the data as well as contain it.
For example, execution counts, is a list of 3-tuples, `name, count, miss`. But it also needs data on how to present it:
It should be sorted by count, have a ratio and cumulative ratio column, and the miss should be presented as a percentage.
That could be a method, which converts the data to a table (where table is `list[tuple[str]]`)

This PR contains the following comment:
> A Table defines how to convert a set of Stats into a specific set of rows displaying some aspect of the data.

That should be a method on the Stats (or Section):
```Py
Rows: TypeAlias = list[tuple[str]]
def to_rows(self) -> Rows:
     ...
```

**Comment by mdboom** [comment]
> I think what bothers me with this PR is that the data processing is mixed in with the data storage. 

- raw data (Stats)
- abstract views of that data (Table)
- organization of that data (Section)
- output (output_markdown)

> There is nothing wrong with objects that enhance and structure the data, but data processing should be separated, otherwise the code can be hard to follow.

I agree 100%, but I think this refactor does that.



def main():
    # Process args and find folders, etc.
    raw_stats = gather_stats(stats_dir)
    stats = structure_stats(raw_stats)
    output_stats(stats, outfile)

> The Stats class should be a structured data type, with the top level containing attributes for each of the top level categories in the data, execution_counts, pair_counts, predecessor_pairs, etc.

> It could simply be a list of Sections, where the Sections describe the data, if that makes more sense. The markdown file is structured as a list of sections.

Wouldn't that be *more* of a combining of processing and presentation?

I think what would address your concerns is:

- The `calc_*_` functions become methods on the `Stats` class.  
- The `Table` class would go away.
- The `Section` class would describe the organization of the file and how tables need to be combined.

**Comment by mdboom** [comment]
@markshannon: I've largely moved the data processing inside of the `Stats` class and the new `OpcodeStats` class.  The `Table`/`Section` distinction is still required, since a Section may have multiple tables etc.  But I hope this is closer to what you had in mind in terms of separation of concerns.

**Comment by markshannon** [comment]
Have you checked that the latest commit produces the same output as main?

**Comment by mdboom** [comment]
> Have you checked that the latest commit produces the same output as main?

Yes, for single datasets.  For comparative the results are different, but due to bugfixes.
